### ISSUE TITLE
Display datetimes in the user's system timezone

### ISSUE THREAD

**Issue by rupalkhilari** [electology/approval_frame#88]
The default TIME_ZONE as set in the settings file is 'America/Los_Angeles' resulting in all dates and times being rendered in this particular timezone (irrespective of the user's timezone) . (Thanks @kushankr  for pointing this out!) . Thus the dates and times displayed in the home page can be quite confusing to the user.

We would like all dates and times to be rendered in the user's timezone automatically without having the user select his/her timezone during registration. As extensively discussed in https://github.com/electology/approval_frame/issues/75 we might need to use some client side Javascript library (like jstz) to help us set the app TIME_ZONE to the user's system timezone. Also, as recommended widely - setting TIME_ZONE='UTC' in settings.py. In addition, also add the appropriate timezone abbreviation (PST, IST, EST) whenever displaying datetimes.

**Comment by kushankr** [comment]
Since this was initially a part of #75, I will be working on this! Thanks Rupal for raising this as a separate issue!

**Comment by rupalkhilari** [comment]
Thanks Kushank !

**Comment by kushankr** [comment]
@rupalkhilari 

Hey Rupal,
"We would like all dates and times to be rendered in the user's timezone automatically" -> Regarding this, should the timezone of user be activated on login or should it be activated by default?
Here is the thing:
I am loading a javascript for 'base.html'. Through this js code, I want to send user's timezone obtained by 'jstz' to view and save it. 

There are 2 ways i have implemented this:

1) AJAX

2) Append a form and a hidden input and submit form.

1-> AJAX -> A big advantage is that sending timezone data is asynchronous, so it will not affect the loading time of pages dependent on 'base.html'. Normal post does affect the loading time of pages. AJAX serves us well, except ->  When the user loads 'index.html' for the first time, the timezone is UTC. Subsequent reloads and transitions to other pages (such as my-polls page) reflect the correct user timezone. As an example, I was working with AJAX and everything was fine but when i deleted my browser history and reloaded the 'approval_polls' url, it showed date times in UTC.  When I refreshed the page or clicked 'Home' again, the date times corresponding to user timezone were correctly displayed. 
I have tried several options (async: false is deprecated) but there seems to be no way around this.

2->NON AJAX POST -> In this method, I am appending a form and a hidden input and submitting the form through a .js file. One major problem is that it affects the loading time of pages reliant on 'base.html'. The delay effect due to adding post request is 1-2 sec (since it is synchronous). Also, since i am passing timezone data to a view, it expects a response object and when i redirect to the referring page (user should be kept on the same page), once again 'base.html' loads (as all templates inherit from it) and the whole thing runs over and over. I have used 'document.cookie' in javascript to overcome this problem (and ensure form submission happens only once) but then i thought to myself 'what if user's cookies are disabled'. 


Can you please share your thoughts on this? Should I not worry about user's cookies being disabled and use NON AJAX POST? Is NON AJAX POST the way should i be going?

I would prefer to use AJAX (because it was recommended widely and it has no effect on loading time of pages) but to use AJAX, i might have to make certain modifications:
We can display UTC/GMT times (with abbrv) when user is not authenticated and display user's local timezone times (with abbrv) when user is authenticated.

I am sorry for this huge post. I would appreciate your feedback.

**Comment by rupalkhilari** [comment]
@kushankr Thanks for the detailed post. I haven't worked on such a problem earlier so it was a good insight for me too. 
I too would go for the AJAX route, since i see how the synchronous form post methods could add a significant delay. Also we would want to avoid using cookies at this point which might fail in case they have been disabled. 
As I understand it, I am assuming that the sequence of events in (1 - AJAX approach) is like this. Please correct me if something is not right:

1) base.html finds out the timezone name through jstz, and sends a GET request using AJAX to the set the timezone. 
2) The middleware intercepts the request and attempts to activate the user's timezone (as you mentioned this would not happen at the first request due to its async nature, hence the dates and times will be displayed in UTC in the template the first time around)
3) The view sets the timezone in a request.session variable.
4) On the second request to the same page, the user timezone will be activated in the middleware and hence the local timezone dates will be displayed. 

This seems good, but the point about which I am concerned is that it is possible that the user might embed the poll in a website which might contain the countdown to the poll and the date published. In such a case would it be possible to guarantee that the user sees the time in the local timezone and countdown the first time? 

Do you think it will make sense to further modify the AJAX request to reload the page on the set-user-timezone success ? We could prevent the recursive loading of pages by checking if the current timezone is that as found by jstz, if so not reload the page again. That could help prevent sending the AJAX request each time the base.html is loaded, although jstz will compute the timezone each time to check if there were any changes on the system. Perhaps something like this in base.html:

    {% load tz %}
    {% get_current_timezone as TIME_ZONE %}
    <script>
             var timezone = jstz.determine();
             var timezone_name = timezone.name();
              if (timezone_name != "{{ TIME_ZONE }}") {
              $.ajax({
                  url : '/approval_polls/set-user-timezone/',
                  action : "GET",
                  data : {
                    timezone : timezone_name
                  },
                  success : function(data){
                     if (data.status == "success") {
                      // alert("Setting timezone to " + timezone_name); 
                      window.location.reload(true);
                    }
                  },
                  error : function(data){
                      console.log("There was an error");
                  }
              })
            }
    </script>

Since the default session would last 2 weeks from the last accessed time, maybe later on we could store and update these timezones against the usernames (when the users are logged in) - but it may not be a priority for now. For now these reloads should happen only once whenever the system timezone has been changed or when the session expires. Please let me know what you think.

**Comment by kushankr** [comment]
Thank you, Rupal for your reply.  Your understanding is correct. I found your suggestions very valuable.

The AJAX code you posted is EXACTLY what I wrote too! Recursive loading of pages was a problem in NON AJAX POST. In AJAX, the pages do not re load.

I have tried this: window.location.reload(); on success before. I had a concern that the date time changes as the user is looking at it.  So if i am loading /approval_polls/ for the first time, it will first  show UTC and refreshes the page to show time in user's timezone. As an example, it shows 5:15 P.M. UTC and changes within a second to 12:15 PM. EST. I wanted to avoid user seeing the transition and only see 12:15 PM. EST for the first time . But if you are fine with it, we can go with this ajax code. 

OR Since the page is already reloading the first time, we can combine ajax and non-ajax post to prevent user from seeing the transition. Non-ajax post for the first time will carry a small time penalty (1-2 sec) but it is no different than ajax refreshing page on success.

Here is the code: Please let me know what you think. 

    <script>
      var timezone = jstz.determine();
      var timezone_name = timezone.name();
      if (timezone_name == "{{ TIME_ZONE }}") {
        $.ajax({
          method: 'GET',
          url: '/approval_polls/set-user-timezone/',
          data: {'tzone': timezone_name},
          success: function (data) {
          },
          error: function (data) {
                 alert("it didnt work");
          }
        });
      }
      else {
        $('body').append('<form id="set_timezone"></form>');
        $('#set_timezone') 
              .attr("action","/approval_polls/set-user-timezone/") .attr("method","get")
              .append('<input type="hidden" name="tzone" value='+timezone_name+'>')
        $('#set_timezone').submit()
      }      
    </script>

**Comment by rupalkhilari** [comment]
Hi Kushank, 

Sure, thats a good suggestion to prevent the reloading. Please go ahead with this and we can test it out further once you submit the PR. 

    if (timezone_name == "{{ TIME_ZONE }}") {

will this be sending GET requests even when the apps current timezone and user's timezone are the same ?

**Comment by kushankr** [comment]
Hi Rupal,

I have modified the code since morning as I was thinking over it. Yes, you are right that with the script i provided,  AJAX requests will be sent even when apps current timezone and user's timezone are the same. I am still modifying the script. 

I will keep you updated.

**Comment by kushankr** [comment]
Here is the modified script:

    <script>
      var timezone = jstz.determine();
      var timezone_name = timezone.name();
      if (timezone_name != "{{ TIME_ZONE }}") {
        $('body').append('<form id="set_timezone"></form>');
        $('#set_timezone') 
              .attr("action","/approval_polls/set-user-timezone/") .attr("method","get")
              .append('<input type="hidden" name="timezone" value='+timezone_name+'>')
        $('#set_timezone').submit()
      }  
    </script>

I have tested this, this NON AJAX method is only called when user's timezone and app's timezone are different. This happens usually once in the app -> when a user loads any page for the first time.  After this, user's timezone is accurately set and this is not called again. AJAX might not be needed.

**Comment by rupalkhilari** [comment]
@kushankr , thats good. Hopefully all should work well with this non Ajax route.

Thanks!

**Comment by kushankr** [comment]
Initiated a PR.
#89 

Thank you, Rupal, for helping me out with this issue. When you get time, please review my PR and let me know what you think. Thanks a lot, Rupal.

**Comment by rupalkhilari** [comment]
Thank you Kushank, also for patiently working to find the optimum solution to this. I checked out the changes, everything seemed all good to merge except a couple of very minor things.

**Comment by kushankr** [comment]
Thank YOU, Rupal. Without your input and suggestions, I would not have been able to find the optimal solution. Thank you for helping me out with this issue. 

I have updated the PR.

**Comment by rupalkhilari** [comment]
https://github.com/electology/approval_frame/pull/89 
Thanks @kushankr ! I too learnt new things on the way !
